src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java

```diff
@@ -1,85 +1,110 @@
 
     public FeederImpl(){
         super(1, TimeUnit.SECONDS);
         
-        this.beltController = new WPI_TalonSRX(RobotMap.BELT_TALON_PORT);
-        this.hopperController = new WPI_TalonSRX(RobotMap.HOPPER_TALON_PORT);
-        this.feederUltrasonic = new Ultrasonic(RobotMap.ULTRASONIC_PORTS[0], RobotMap.ULTRASONIC_PORTS[1]);
+        this.topUltrasonic = new Ultrasonic(RobotMap.ULTRASONIC_PORTS[0], RobotMap.ULTRASONIC_PORTS[1]);
         
+        this.conveyorSubsystem = new ConveyorImpl();
+        this.hopperSubsystem = new HopperImpl();
 
-        this.poller = new UltrasonicPoller(feederUltrasonic, 3000, 300);
-        this.distanceTuner = new SmartDashboardTuner("Ball Distance", 300, d -> this.poller.setDistance(d));
-        
-        this.hopperController.configContinuousCurrentLimit(20, 250);
-        this.hopperController.enableCurrentLimit(false);
-    }
+        this.topUltrasonicDistanceMM = 30;
+        this.topUltrasonicTuner = new SmartDashboardTuner("Top Ultrasonic Tuner", this.topUltrasonicDistanceMM, x -> this.topUltrasonicDistanceMM = x);
 
-    public void enableConveyor() {
-        this.beltController.set(-0.95d);
-        // TODO Find actual values
-    }
+        this.topPoller = new UltrasonicPoller(this.topUltrasonic, 3000, 20); // Temp values
 
-    public void enableHopper() {
-        this.hopperController.enableCurrentLimit(true);
-        this.hopperController.set(0.8d);
     }
 
-    public void stopConveyor() {
-        this.beltController.set(0.0d);
+    public void task() {
+        // PLACEHOLDER
+        return;
+    }
 
+    @Override
+    public void defineResources() {
+        require(this.topUltrasonic);
     }
 
-    public void stopHopper() {
-        this.hopperController.enableCurrentLimit(false);
-        this.hopperController.set(0.0d);
+    @Override
+    public void prime() {
+        if(this.topUltrasonic.getRangeMM() > this.topUltrasonicDistanceMM) {
+            this.enable(FeederComponent.HOPPER);
+            this.enable(FeederComponent.CONVEYOR);
+        } else {
+            this.stop(FeederComponent.HOPPER);
+            this.stop(FeederComponent.CONVEYOR);
+        }
     }
 
-    public void stopAll() {
-        //this.beltController.set(0.0d);
-        //this.hopperController.set(0.0d);
-        this.stopHopper();
-        this.stopConveyor();
+    @Override
+    public void unload() {
+        if(!this.topPoller.noBalls) {
+            this.enable(FeederComponent.HOPPER);
+            this.enable(FeederComponent.CONVEYOR);
+        } else {
+            this.stop(FeederComponent.HOPPER);
+            this.stop(FeederComponent.CONVEYOR);
+        }
     }
 
-    public void reverseFeeder() {
-        this.beltController.set(0.6d);
-        this.hopperController.set(-0.8d);
-        // TODO see above
+    @Override
+    public void enable(FeederComponent component, FeederDirection direction) {
+        // Probably shouldn't have a default case IMO
```

Comment:
On the contrary, IMO *every* switch statement should have a default, even when it's a switch over an enum and you mean to have a defined behaviour for each case. In that scenario the best practice is to add a default case which throws an exception, so you don't accidentally add a new enum value in the future and then get weird unexpected runtime behaviour and bugs because it wasn't accounted for. Having it fall into a default exception case means you're going to notice it much sooner and get around to implementing the new behaviour without falling into the "wtf is happening" debug trap.